Alembic Migrations

Whiteknight07 opened the pull request:
## 📝 Description

Introduces Alembic for database schema management, eliminating need to manually delete `artifactminer.db` on schema changes. Includes initial migration capturing Milestone 1 schema + new models for skills, resume items, and exports.

  **Key changes:**
  - Alembic config & environment setup
  - Initial migration: 11 tables (artifacts, questions, consents, repo_stats, user_repo_stats, user_intelligence_summaries, user_answers, uploaded_zips, skills, project_skills, resume_items, exports)
  - Extended existing models: added `frameworks`, `collaboration_metadata`, `ranking_score`, `deleted_at` to `RepoStat`; added `activity_breakdown` to `UserRepoStat`; added `extraction_path` to `UploadedZip`
  - Foreign key cascades and bidirectional relationships on new models
  - Unique constraint on `ProjectSkill(repo_stat_id, skill_id)` to prevent duplicates
  - README docs for migration workflow
  
 **Note**: Most of the files were autogenerated through the `alembic init` command.  I have only created new models in `models.py` as they are needed for milestone 1 and added docs in `README.md`. 
  
  **Closes:** #166 

  ---

  ## 🔧 Type of Change

  - [x] ✨ New feature (non-breaking change that adds functionality)
  - [x] 📚 Documentation added/updated
  - [x] ♻️ Refactoring
   Please check `README.md` on how to use alembic. 
  ---

  ## 🧪 Testing

  **Tested locally:**
  - [x] Fresh DB creation: `uv run alembic upgrade head` on empty dir
  - [x] Migration idempotency: re-running upgrade produces no changes
  - [x] Downgrade/upgrade cycle: `alembic downgrade -1 && alembic upgrade head` works as expected.

  **Manual verification:**
  - [x] All tables created with correct columns, indexes, and constraints
  - [x] Foreign key cascades work (deleting repo_stat removes related project_skills/resume_items)
  - [x] Unique constraint prevents duplicate ProjectSkill entries

  ---

  ## ✓ Checklist

  - [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
  - [x] 💬 I have commented my code where needed
  - [x] 📖 I have made corresponding changes to the documentation
  - [x] ⚠️ My changes generate no new warnings
  - [ ] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
  - [x] 🔗 Any dependent changes have been merged and published in downstream modules
  - [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

  ---

shahshlok commented on alembic/versions/b30abc89d0a8_add_cascade_and_relationships_to_new_.py:
```diff
@@ -0,0 +1,44 @@
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_unique_constraint('uq_project_skill', 'project_skills', ['repo_stat_id', 'skill_id'])
+    op.drop_constraint(None, 'project_skills', type_='foreignkey')
+    op.drop_constraint(None, 'project_skills', type_='foreignkey')
```
A lot of the Alembic commands in both the `upgrade()` and `downgrade()` use `None` instead of actual constraint names such as the one that I highlighted. Alembic won't know which constraints to drop.

Second, having 2 identical commands will result in an error or warning. 

Please make replace `None` with something more descriptive and make sure that a single command is declared only once

Whiteknight07 commented:
Update: I reworked the migrations to handle SQLite cleanly and still satisfy the constraint requirements.

  - Moved the project_skills changes into the initial migration (3b3f297185db_add_milestone_1_schema.py): the table now defines ondelete='CASCADE' on both FKs and a uq_project_skill unique constraint directly at creation time.
  - Since we only use SQLite and project_skills was empty, this is safe and keeps the actual SQLite schema in sync with the SQLAlchemy models.
  - The follow-up migration (b30abc89d0a8_add_cascade_and_relationships_to_new_.py) is now a no-op (upgrade/downgrade just pass), because its original intent is fully handled in the first migration.

  This removes the None constraint names, avoids duplicate drop_constraint calls, and ensures the uniqueness and cascade behavior are actually enforced in our SQLite database.
